specifyweb/specify/calculated_fields.py

```diff
@@ -15,98 +14,113 @@ def get_model(name: str):
     """
     return getattr(models, name.capitalize())
 
+def calculate_quantity(obj, field_name: str) -> int:
+    return obj.aggregate(total=Sum(field_name))["total"] or 0
+
+def calculate_actual_count(obj, preparations):
+    actualTotalCountAmt = 0
+    for prep in preparations:
+        giftpreparation_quantity = calculate_quantity(prep.giftpreparations, "quantity")
+        exchangeoutprep_quantity = calculate_quantity(prep.exchangeoutpreps, "quantity")
+        disposalpreparation_quantity = calculate_quantity(prep.disposalpreparations, "quantity")
+
+        countamt = obj.countamt or 0
+        available = max(0, countamt - giftpreparation_quantity - exchangeoutprep_quantity - disposalpreparation_quantity)
+        actualTotalCountAmt += available
+    return actualTotalCountAmt
+
+def calculate_totals(obj, model_name):
+    model = get_model(model_name)
+    totalPreps = model.objects.filter(deaccession=obj).count()
+    totalItems = model.objects.filter(deaccession=obj).aggregate(total=Sum("totalItems"))["total"] or 0
+    return totalPreps, totalItems
```

Comment:
This doesn't work when a disposal, gift or exchangeout object doesn't have a deaccession record linked to it. Are disposals, gifts and exchangeouts guaranteed to have a deaccession linked to them?

Reply:
Not at a database level. At least for gift. Check others as an excercise.
output of `show create table gift`. On sp7demofish, I am able to save a gift without deaccession so doesn't look like there is a business rule either: https://sp7demofish.specifycloud.org/specify/view/gift/479/
```
CREATE TABLE `gift` (
  `GiftID` int(11) NOT NULL AUTO_INCREMENT,
  `TimestampCreated` datetime NOT NULL,
  `TimestampModified` datetime DEFAULT NULL,
  `Version` int(11) DEFAULT NULL,
  `Contents` text DEFAULT NULL,
  `Date1` date DEFAULT NULL,
  `Date1Precision` tinyint(4) DEFAULT NULL,
  `DateReceived` date DEFAULT NULL,
  `GiftDate` date DEFAULT NULL,
  `GiftNumber` varchar(50) NOT NULL,
  `Integer1` int(11) DEFAULT NULL,
  `Integer2` int(11) DEFAULT NULL,
  `Integer3` int(11) DEFAULT NULL,
  `IsFinancialResponsibility` bit(1) DEFAULT NULL,
  `Number1` decimal(20,10) DEFAULT NULL,
  `Number2` decimal(20,10) DEFAULT NULL,
  `PurposeOfGift` varchar(64) DEFAULT NULL,
  `ReceivedComments` varchar(255) DEFAULT NULL,
  `Remarks` text DEFAULT NULL,
  `SpecialConditions` text DEFAULT NULL,
  `SrcGeography` varchar(500) DEFAULT NULL,
  `SrcTaxonomy` varchar(500) DEFAULT NULL,
  `Status` varchar(64) DEFAULT NULL,
  `Text1` text DEFAULT NULL,
  `Text2` text DEFAULT NULL,
  `Text3` text DEFAULT NULL,
  `Text4` text DEFAULT NULL,
  `Text5` varchar(128) DEFAULT NULL,
  `YesNo1` bit(1) DEFAULT NULL,
  `YesNo2` bit(1) DEFAULT NULL,
  `CreatedByAgentID` int(11) DEFAULT NULL,
  `DisciplineID` int(11) NOT NULL,
  `AddressOfRecordID` int(11) DEFAULT NULL,
  `DivisionID` int(11) DEFAULT NULL,
  `ModifiedByAgentID` int(11) DEFAULT NULL,
  `DeaccessionID` int(11) DEFAULT NULL,
  PRIMARY KEY (`GiftID`),
  KEY `GiftDateIDX` (`GiftDate`),
  KEY `GiftNumberIDX` (`GiftNumber`),
  KEY `FK3069307699B003` (`CreatedByAgentID`),
  KEY `FK306930DC8B4810` (`AddressOfRecordID`),
  KEY `FK306930BE26B05E` (`DeaccessionID`),
  KEY `FK3069304CE675DE` (`DisciplineID`),
  KEY `FK30693097C961D8` (`DivisionID`),
  KEY `FK3069305327F942` (`ModifiedByAgentID`),
  CONSTRAINT `FK3069304CE675DE` FOREIGN KEY (`DisciplineID`) REFERENCES `discipline` (`UserGroupScopeId`),
  CONSTRAINT `FK3069305327F942` FOREIGN KEY (`ModifiedByAgentID`) REFERENCES `agent` (`AgentID`),
  CONSTRAINT `FK3069307699B003` FOREIGN KEY (`CreatedByAgentID`) REFERENCES `agent` (`AgentID`),
  CONSTRAINT `FK30693097C961D8` FOREIGN KEY (`DivisionID`) REFERENCES `division` (`UserGroupScopeId`),
  CONSTRAINT `FK306930BE26B05E` FOREIGN KEY (`DeaccessionID`) REFERENCES `deaccession` (`DeaccessionID`),
  CONSTRAINT `FK306930DC8B4810` FOREIGN KEY (`AddressOfRecordID`) REFERENCES `addressofrecord` (`AddressOfRecordID`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_general_ci
```

Reply:
The API and the frontend break for me for all 3
<img width="804" alt="image" src="https://github.com/specify/specify7/assets/16229739/fcc2bf12-5f29-4ad2-a727-c9544180fa5e">

Reply:
lol debug it then